pkg/statistics/handle/usage/indexusage/collector.go

```diff
@@ -0,0 +1,225 @@
+	item := m[id]
+	item.QueryTotal += sample.QueryTotal
+	item.RowAccessTotal += sample.RowAccessTotal
+	item.KvReqTotal += sample.KvReqTotal
+	for i, val := range sample.PercentageAccess {
+		item.PercentageAccess[i] += val
+	}
+
+	if item.LastUsedAt.Before(sample.LastUsedAt) {
+		item.LastUsedAt = sample.LastUsedAt
+	}
+	m[id] = item
+}
+
+func (m indexUsage) update(tableID int64, indexID int64, sample Sample) {
+	id := GlobalIndexID{TableID: tableID, IndexID: indexID}
+	m.updateByKey(id, sample)
+}
+
+func (m indexUsage) merge(destMap indexUsage) {
+	for id := range destMap {
+		item := destMap[id]
+		m.updateByKey(id, item)
+	}
+}
+
+// Collector represents a data structure to record the index usage for the whole node
+type Collector struct {
+	collector collector.GlobalCollector[indexUsage]
+	mapper    indexUsage
+	sync.RWMutex
+
+	closed bool
+}
+
+// NewCollector create an index usage collector
+func NewCollector() *Collector {
+	iuc := &Collector{
+		mapper: make(indexUsage),
+	}
+	iuc.collector = collector.NewGlobalCollector[indexUsage](iuc.merge)
+
+	return iuc
+}
+
+// GetIndexUsage returns the index usage information
+func (c *Collector) GetIndexUsage(tableID int64, indexID int64) *Sample {
+	c.RLock()
+	defer c.RUnlock()
+
+	info, ok := c.mapper[GlobalIndexID{tableID, indexID}]
+	if !ok {
+		return nil
+	}
+	return &info
+}
+
+func (c *Collector) merge(delta indexUsage) {
+	c.Lock()
+	defer c.Unlock()
+
+	c.mapper.merge(delta)
+}
+
+// SpawnSessionCollector creates a new session collector attached to this global collector
+func (c *Collector) SpawnSessionCollector() *SessionIndexUsageCollector {
+	return &SessionIndexUsageCollector{
+		mapper:    make(indexUsage),
+		collector: c.collector.SpawnSession(),
+	}
+}
+
+// StartWorker starts the background worker inside
+func (c *Collector) StartWorker() {
+	c.collector.StartWorker()
+}
+
+// Close closes the background worker inside
+func (c *Collector) Close() {
+	c.collector.Close()
+}
+
+// GCIndexUsage will delete the usage information of non-existent indexes.
+// `tableMetaLookup` argument is represented as a function (but not `sessionctx.Context` or `infoschema.InfoSchema`) to
+// avoid depending on `sessionctx.Context`.
+func (c *Collector) GCIndexUsage(tableMetaLookup func(id int64) (*model.TableInfo, bool)) {
+	// it's possible to split `s.Mutex` into multiple mutex to avoid blocking the creation of session and sweepiing index
+	// However, as all these operations are infrequent, keeping a simpler mutex is enough.
+	c.Lock()
+	defer c.Unlock()
+	for k := range c.mapper {
+		tbl, ok := tableMetaLookup(k.TableID)
+		if !ok {
+			delete(c.mapper, k)
+			continue
+		}
+		foundIdx := false
+		for _, idx := range tbl.Indices {
+			if idx.ID == k.IndexID {
+				foundIdx = true
+				break
+			}
+		}
+		if !foundIdx {
+			delete(c.mapper, k)
+		}
+	}
+}
+
+// SessionIndexUsageCollector collects index usage per-session
+type SessionIndexUsageCollector struct {
+	mapper indexUsage
```

Comment:
`mapper` seems too general and not a good name, could we rename it?

Reply:
Yes! It has been renamed to `indexUsage`.